### PR TITLE
Arbitrary committee change

### PR DESCRIPTION
**Pull request by asonnino**
Allows to change the committee at any point, even if it is not the end of an epoch. This is useful if we wants to change the committee to only update some network addresses.

### CODE REVIEW

**Review by akichidis — COMMENTED** [review]
Thanks @asonnino for this! A few things from my side:

1. Could you please add tests to explicitly test the clean up/ non clean up behaviour of the components that consume the new committee
2. What about the [StateHandler](https://github.com/MystenLabs/narwhal/blob/main/primary/src/state_handler.rs#L98)? Should we also make sure that consensus round is not getting zeroed when epoch doesn't change? I know this is coming from another input channel but still we might need to check that. 
3. As I wrote on my comment I am a bit skeptical on handling the committee update (within same epoch) and new committee (new epoch) via the same message. I would look into split those - maybe not in this PR but in the future.

**Comment by asonnino** [comment]
@akichidis  I think you are right, it seems better practice to handle the committee update in a separate message. Let me do that

File: consensus/src/consensus.rs [hunk 1]
```diff
@@ -243,15 +243,19 @@ where
         })
     }
 
-    fn reconfigure(&mut self, new_committee: Committee) -> StoreResult<ConsensusState> {
+    fn reconfigure(&mut self, new_committee: Committee) -> StoreResult<Option<ConsensusState>> {
+        let old_epoch = self.committee.epoch();
         self.committee = new_committee.clone();
         self.protocol.update_committee(new_committee)?;
```
[the remark this answers is not in the document]
Comment (asonnino) [reply consensus/src/consensus.rs:248]: fixed

File: primary/src/core.rs [hunk 2]
```diff
@@ -526,15 +526,18 @@ impl Core {
...
-        self.processing.clear();
-        self.certificates_aggregators.clear();
-        self.cancel_handlers.clear();
-
+        let old_epoch = self.committee.epoch();
         self.committee = committee;
-        self.synchronizer.update_genesis(&self.committee);
         tracing::debug!("Committee updated to {}", self.committee);
+
+        if self.committee.epoch() > old_epoch {
```
[the remark this answers is not in the document]
Comment (asonnino) [reply primary/src/core.rs]: fixed

File: primary/src/certificate_waiter.rs [hunk 3]
```diff
@@ -160,9 +160,13 @@ impl CertificateWaiter {
...
                     match message {
                         ReconfigureNotification::NewCommittee(committee) => {
+                            let old_epoch = self.committee.epoch();
                             self.committee = committee;
                             tracing::debug!("Committee updated to {}", self.committee);
-                            self.pending.clear();
+
+                            if self.committee.epoch() > old_epoch {
+                                self.pending.clear();
+                            }
```
[the remark this answers is not in the document]
Comment (asonnino) [reply primary/src/certificate_waiter.rs]: done

File: consensus/src/consensus.rs [hunk 4]
```diff
@@ -243,15 +243,19 @@ where
...
         tracing::debug!("Committee updated to {}", self.committee);
 
-        self.consensus_index = 0;
+        if self.committee.epoch() > old_epoch {
```
[the remark this answers is not in the document]
Comment (asonnino) [reply consensus/src/consensus.rs]: done

File: primary/tests/epoch_change.rs [hunk 5]
```diff
@@ -421,3 +421,115 @@ async fn test_restart_with_new_committee_change() {
...
+                    config::utils::get_available_port()
+                )
+                .parse()
+                .unwrap();
+
+                total_stake += authority.stake;
+            }
+        }
+
+        // Notify the old committee to change epoch.
```
Comment (akichidis) [inline primary/tests/epoch_change.rs:504]: `Notify the old committee to change epoch.` --> we don't want to change epoch right? Just update the existing committee

**Review by akichidis — APPROVED** [review]
Awesome work, thanks so much @asonnino for this! 💯

File: types/src/primary.rs [hunk 6]
```diff
@@ -545,11 +545,13 @@ pub enum PrimaryMessage {
     },
 }
 
-/// Message to reconfigure worker tasks.
+/// Message to reconfigure worker tasks. This message must be sent by a trusted source.
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub enum ReconfigureNotification {
-    /// Indicate the committee has been updated.
+    /// Indicate the committee has changed. This happens at epoch change.
     NewCommittee(Committee),
```
[the remark this answers is not in the document]
Comment (asonnino) [reply types/src/primary.rs:552]: Good point, let me rename NewCommittee -> NewEpoch

File: primary/tests/epoch_change.rs (hunk 5 above)
Comment (asonnino) [reply primary/tests/epoch_change.rs:504]: good catch, fixing

**Pull request merged**